lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+        let pci_port = PciVirtioSoftNpuPort::new(
+            Self::generate_mac(),
+            data_handles.clone(),
+            virtio.clone(),
+            pipeline.clone(),
+            log.clone(),
+        );
+
+        Ok(Arc::new(SoftNpu {
+            data_links,
+            data_handles,
+            virtio,
+            pci_port,
+            uart,
+            p9fs,
+            pipeline,
+            booted: Mutex::new(false),
+            log,
+        }))
+    }
+
+    /// Generate a mac address with the Oxide OUI for the leading bits and then
+    /// something random in the range of 0xf00000 - 0xf00000 per RFD 174.
+    fn generate_mac() -> [u8; 6] {
+        let mut rng = rand::thread_rng();
+        let m = rng.gen_range::<u32, _>(0xf00000..0xffffff).to_le_bytes();
+        [0xa8, 0x40, 0x25, m[0], m[1], m[2]]
+    }
+
+    /// Set up a dlpi handle for each data link.
+    fn data_handles(data_links: &Vec<String>) -> Result<Vec<dlpi::DlpiHandle>> {
+        let mut handles = Vec::new();
+        for x in data_links {
+            let h = dlpi::open(x, dlpi::sys::DLPI_RAW)?;
+
+            // Although we bind to the IPv6 SAP (Ethertype), the DL_PROMISC_SAP
+            // allows us to pick up everything. Binding to *something* to start
+            // with appears to be required to get packets.
+            dlpi::bind(h, 0x86dd)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_MULTI)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_SAP)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_PHYS)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_RX_ONLY)?;
+            handles.push(h);
+        }
+        Ok(handles)
+    }
+
+    /// Start the management handler for servicing requests from the guest over
+    /// the provided uart device.
+    fn run_management_handler_thread(&self) {
+        info!(self.log, "softnpu: running management handler");
+        self.uart.set_autodiscard(false);
+
+        let log = self.log.clone();
+        let uart = self.uart.clone();
+        let pipeline = self.pipeline.clone();
+        let radix = self.data_links.len();
+
+        spawn(move || Self::management_handler(uart, pipeline, radix, log));
+    }
+
+    fn management_handler(
+        uart: Arc<LpcUart>,
+        pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
+        radix: usize,
+        log: Logger,
+    ) {
+        info!(log, "management handler thread started");
+        loop {
+            let r = ManagementMessageReader::new(uart.clone(), log.clone());
+            let msg = r.read();
+            info!(log, "received management message: {:#?}", msg);
+
+            let pipeline = pipeline.clone();
+            let uart = uart.clone();
+            let log = log.clone();
+            handle_management_message(msg, pipeline, uart, radix, log.clone());
+            info!(log, "handled management message");
+        }
+    }
+}
+
+impl Entity for SoftNpu {
```

Comment:
Are SoftNpu and/or P9 devices expected to be migratable? If not they should also implement `migrate` and return `Migrator::NonMigratable`. (If they are then you probably need some instance spec changes; left a comment over in that module describing those.)